### PR TITLE
mds: limit the snapshot names to 240 characters

### PR DESCRIPTION
**Pull request by luis-henrix**
This PR comes after [the discussion in this thread](https://lore.kernel.org/all/20220305122527.1102109-1-xiubli@redhat.com/).  Note that this PR requires PR #45192 to be merged as well (probably this dependency can be specified in github...).

Basically, without limting snapshot names, the long snapshot names (those starting with '_') will exceed the maximum allowed file names size.

I'd like to get reviews from, at least, @lxbsz and @jtlayton 

Signed-off-by: Luís Henriques <lhenriques@suse.de>

### CODE REVIEW

File: doc/dev/cephfs-snapshots.rst
Comment (lxbsz) [inline doc/dev/cephfs-snapshots.rst:61]: It's worth to explain why ?

File: doc/dev/cephfs-snapshots.rst
Comment (lxbsz) [inline doc/dev/cephfs-snapshots.rst:57]: s/as these are/as it's/ ?

**Comment by luis-henrix** [comment]
Thanks @lxbsz I've pushed a new revision that should address both comments.

File: doc/dev/cephfs-snapshots.rst
Comment (lxbsz) [inline doc/dev/cephfs-snapshots.rst:63]: We'd better use the same format example as `_<SNAPSHOT-NAME>_<INODE-NUMBER>`. 
The others look good to me.

File: doc/dev/cephfs-snapshots.rst
Comment (jtlayton) [inline doc/dev/cephfs-snapshots.rst:59]: Ok, so I guess the client will need to prepend these with a different character?

**Review by jtlayton — COMMENTED** [review]
Looks sane overall, but I'll wait for the doc revision before I approve.

File: doc/dev/cephfs-snapshots.rst
Comment (luis-henrix) [reply doc/dev/cephfs-snapshots.rst:63]: Yeah, that looks better.  I'll update my branch in a bit with this change.  Thanks @lxbsz

File: doc/dev/cephfs-snapshots.rst
Comment (luis-henrix) [reply doc/dev/cephfs-snapshots.rst:59]: I'm not sure I understand your question.  These names that start with an underscore are generated by the MDS itself (although I **really** have no idea about how snapshots work and how these special inodes are used internally).

If you're talking about the encrypted snapshot names, which will be base64-encoded strings that may start with the '_' character, prepending another character is probably the best solution.  But TBH I didn't thought a lot about it, as I've been having fun playing whack-a-mole with the MDS snapshots name handling :-)

Comment (luis-henrix) [reply doc/dev/cephfs-snapshots.rst:59]: > Then #45208 makes no sense any more, and I will close it.

Well, what I wanted with this note was simply to document the current behaviour, i.e. that a user can not do a `mkdir .snap/_my-snapshot`.  The MDS assumes that snapshots starting with an '_' are "special".

PR #45208 isn't changing this behaviour, it's simply adding an extra flag to make things easier for clients to know an inode is a "long name snapshot" (although, I think the client could figure it out anyway comparing the "<INODE-NUMBER>" part of the name with the actual inode->i_ino).  But I may be missing something...?

Comment (lxbsz) [reply doc/dev/cephfs-snapshots.rst:59]: Currently when creating the snapshots if the snapshot name's first char is '_' it will fail anyway even in fscrypt case with the base64 encode names. So in clients side we can figure out whether the snapshot name is a long snap name or not by checking the name[0].

The next step here is that in kclient side how to properly handle the base64 encoded snapshot names, which may start the '_'. When users just create one human readable snapshot whose name without starting the '_', but it still may fail.

File: doc/dev/cephfs-snapshots.rst
Comment (lxbsz) [inline doc/dev/cephfs-snapshots.rst:58]: Maybe we should mention that when creating a snapshot, the snapshot name cannot start with an underscore ('_')... ? If in one `.snap/` when you see the snapshot names starting with '_', they must be inherit from the parent snap realms.

Comment (luis-henrix) [reply doc/dev/cephfs-snapshots.rst:58]: Ok, I'll rephrase to make that clear and to mention the parent snap realms names.  Thanks!

File: doc/dev/cephfs-snapshots.rst
Comment (luis-henrix) [reply doc/dev/cephfs-snapshots.rst:59]: > The next step here is that in kclient side how to properly handle the base64 encoded snapshot names, which may start the '_'. When users just create one human readable snapshot whose name without starting the '_', but it still may fail.

Right, that's what I was going to look at next.  But first I'm still trying to make the client handle correctly these long snapshot names when a directory is encrypted.  I've already an experimental patch, but it still needs some cleanup (and more testing).  Hopefully I'll be able to send this patch later today or tomorrow.

**Review by joscollin — COMMENTED** [review]
@luis-henrix Needs qa test for this.

**Comment by luis-henrix** [comment]
> @luis-henrix Needs qa test for this.

Thanks @joscollin.  I've just pushed another commit that adds a very simple test to verify the MDS is limiting the snapshot names.  I think it's enough to have this test script in that place, and it will be picked by... hmm... something... which will eventually run it.  Or do I need to let that "something" know about the new test?

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (joscollin) [inline qa/workunits/fs/snaps/snaptest-name-limits.sh:38]: @luis-henrix I think we need only the last test (with 241), as the shorter name length test already covered somewhere else? If it errors out, the test is success.

Comment (luis-henrix) [reply qa/workunits/fs/snaps/snaptest-name-limits.sh:38]: Ok, thanks.  I'll remove everything else from the script and push it the new version in a bit.

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (kotreshhr) [inline qa/workunits/fs/snaps/snaptest-name-limits.sh:24]: [ -d d1/.snap/$longname ] && fail "Invalid snapshot exists: $exists" is more readable ?

Comment (luis-henrix) [reply qa/workunits/fs/snaps/snaptest-name-limits.sh]: Thank you @kotreshhr I've just pushed a new revision with your suggestion.

**Review by xiubli — APPROVED** [review]

**Review by kotreshhr — APPROVED** [review]

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (joscollin) [inline qa/workunits/fs/snaps/snaptest-name-limits.sh:8]: drop this, as d1 is removed entirely in the next line?

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (joscollin) [inline qa/workunits/fs/snaps/snaptest-name-limits.sh:22]: `.snap` doesn't exist here. This needs `mkdir -p` to work correctly or create it earlier.

**Review by joscollin — CHANGES_REQUESTED** [review]

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (luis-henrix) [reply qa/workunits/fs/snaps/snaptest-name-limits.sh:22]: Yeah, `.snap` is special here (see [1]).  You don't need to create it as it exists in every cephfs directory.  That's one of the non-POSIX things in cephfs.

[1] https://docs.ceph.com/en/pacific/dev/cephfs-snapshots/

File: qa/workunits/fs/snaps/snaptest-name-limits.sh
Comment (luis-henrix) [reply qa/workunits/fs/snaps/snaptest-name-limits.sh:8]: That would be true for regular directories. `.snap` is special and you can't remove `d1` if you still have snapshots in it.  (See also comment bellow.)

**Review by joscollin — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
Looks good!

File: doc/dev/cephfs-snapshots.rst
Comment (jtlayton) [reply doc/dev/cephfs-snapshots.rst:59]: Yeah, my comment was more about fscrypt. If we have an fscrypted snapshot name, then we'll probably have to prepend it with a legal character.

**Review by jtlayton — APPROVED** [review]

**Comment by vshankar** [comment]
@luis-henrix could you please rebase this please. I'm planning to test out pending fscrypt related mds changes this week.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#08-Dec-2022

**Comment by vshankar** [comment]
jenkins retest this please

File: src/mds/Server.h
Comment (idryomov) [inline src/mds/Server.h:463]: This implies that `<INODE-NUMBER>` (in decimal, as it gets appended) can't be longer than 13 characters.  As far I was able to deduce, this corresponds to 1099511627776 (2^40) which is the cutoff between regular and "system" inodes:
```
bool DataScan::valid_ino(inodeno_t ino) const
{
  return (ino >= inodeno_t((1ull << 40)))
    || (MDS_INO_IS_STRAY(ino))
    || (MDS_INO_IS_MDSDIR(ino))
    || ino == CEPH_INO_ROOT
    || ino == CEPH_INO_CEPH
    || ino == CEPH_INO_LOST_AND_FOUND;
}
```
Was this based on the assumption that `<INODE-NUMBER>` would always be a "system" inode?  If so, it doesn't appear to be the case now (output is courtesy of @batrick):
```
$ ls .snap/
_foo_1099511627776  bar  baz
$ ls c/.snap
_bar_1099511627777  _baz_1099511627777
```
@luis-henrix @lxbsz It's been a while but perhaps you happen to recall some details here?

**Comment by luis-henrix** [comment]
On Tue, Apr 21 2026, Ilya Dryomov wrote:

> @idryomov commented on this pull request.
>
>
>
>>    feature_bitset_t required_client_features;
>  
>    bool forward_all_requests_to_auth = false;
>    bool replay_unsafe_with_closed_session = false;
>    double cap_revoke_eviction_timeout = 0;
>    uint64_t max_snaps_per_dir = 100;
> +  // long snapshot names have the following format: "_<SNAPSHOT-NAME>_<INODE-NUMBER>"
> +  uint64_t snapshot_name_max = NAME_MAX - 1 - 1 - 13;
>
> This implies that `<INODE-NUMBER>` (in decimal, as it gets appended) can't be longer than 13 characters.  As far I was able to deduce, this corresponds to 1099511627776 (2^40) which is the cutoff between "valid" and "system" inodes:
> ```
> bool DataScan::valid_ino(inodeno_t ino) const
> {
>   return (ino >= inodeno_t((1ull << 40)))
>     || (MDS_INO_IS_STRAY(ino))
>     || (MDS_INO_IS_MDSDIR(ino))
>     || ino == CEPH_INO_ROOT
>     || ino == CEPH_INO_CEPH
>     || ino == CEPH_INO_LOST_AND_FOUND;
> }
> ```
> Was this based on the assumption that `<INODE-NUMBER>` would always be a "system" inode?  If so, it doesn't appear to be the case now (output is courtesy of @batrick):
> ```
> $ ls .snap/
> _foo_1099511627776  bar  baz
> $ ls c/.snap
> _bar_1099511627777  _baz_1099511627777
> ```
> @luis-henrix @lxbsz It's been a while but perhaps you happen to recall some details here?

Yikes! That's long gone from my memory.  It's been a long time since last
time I looked into ceph code :-(

You're reasoning above seems to make sense (the 2^40), but my memory can't
bring in any more details.  I had a hard time to even remember about the
long snapshot names.

But what's wrong with the example you're showing above?  It still follows
the 13 chars rule, right?

Cheers,
-- 
Luís

**Comment by idryomov** [comment]
I went through the thread on ceph-devel that is linked in the description of this PR and got even more confused by the snippets that seem to conflict with each other:
```
+/*
+ * The encrypted long snap name will be in format of
+ * "_${ENCRYPTED-LONG-SNAP-NAME}_${INODE-NUM}". And will set the max longth
+ * to sizeof('_') + NAME_MAX + sizeof('_') + max of sizeof(${INO}) + extra 7
+ * bytes to align the total size to 8 bytes.
+ */
+#define CEPH_ENCRYPTED_LONG_SNAP_NAME_MAX (1 + 255 + 1 + 16 + 7)
```
```
'_' + <name> + '_' + '<inode#>' '\0'
 1  +   64   +  1  +    12     +  1 = 80
```
```
For fscrypt case I think it's okay, because the max len of the encrypted name
will be 189 bytes, so even plusing the extra 2 * sizeof('_') - sizeof(<inode#>)
== 15 bytes with ceph PR#41592 it should work well.

But for none fscrypt case, we must limit the max len to NAME_MAX - 2 * 
sizeof('_') - sizeof(<inode#>) == 255 - 2 - 13 == 240.
```
Even keeping in mind the fact that fscrypt support was in the middle of being developed (i.e. very much in flux) at the time, I can't wrap my head around the variance there.  https://github.com/ceph/ceph/pull/45192 didn't change the encoding -- it remained to be decimal -- and yet just a few messages posted in the span of a week suggest that the inode number encoded into a string could take up to 12, 13 or 16 characters?

To me, it seems like 13 characters is the _minimum_ number of characters while the maximum that we are interested in here would be 20 (based on `UINT64_MAX`, i.e. 18446744073709551615).  A couple of constants are reserved by RADOS there but everything else is fair game:
```
/*
 * ino, object, etc.
 */
typedef __le64 ceph_snapid_t;
#define CEPH_SNAPDIR ((__u64)(-1))  /* reserved for hidden .snap dir */
#define CEPH_NOSNAP  ((__u64)(-2))  /* "head", "live" revision */
#define CEPH_MAXSNAP ((__u64)(-3))  /* largest valid snapid */
```

**Comment by idryomov** [comment]
> But what's wrong with the example you're showing above? It still follows the 13 chars rule, right?

It does, but it's a fresh filesystem with just a couple of files/directories/snapshots created on it.  The appended number would only grow from there (1099511627777, 1099511627778, 1099511627779, etc).

@luis-henrix Perhaps you had some kind of an upper limit in mind that still fit in 13 decimal digits?

**Comment by luis-henrix** [comment]
On Wed, Apr 22 2026, Ilya Dryomov wrote:

> *
> idryomov left a comment (ceph/ceph#45312)
>
>  But what's wrong with the example you're showing above? It still follows the 13 chars rule, right?
>
> It does, but it's a fresh filesystem with just a couple of files/directories/snapshots created on it. The appended number would
> only grow from there (1099511627777, 1099511627778, 1099511627779, etc).

Ah! OK, understood.

> @luis-henrix Perhaps you had some kind of an upper limit in mind that still fit in 13 decimal digits?

Unfortunately, I really have no clue why that number was picked.  I've
cloned a fresh copy of ceph user-space to see if anything would ring a
bell, but I couldn't find anything that would justify it.

The best thing I could find was this:

std::string_view SnapInfo::get_long_name() const
{
  if (long_name.empty() ||
      long_name.compare(1, name.size(), name) ||
      long_name.find_last_of("_") != name.size() + 1) {
    std::ostringstream oss;
    oss << "_" << name << "_" << (unsigned long long)ino;
    long_name = oss.str();
  }
  return long_name;
}

which seems to corroborate your point that it should be 20 (for
UINT64_MAX) and not 13.  Maybe Xiubo or Jeff have a better memory than
mine, but it definitely looks like a bug.

Cheers,
-- 
Luís

**Comment by idryomov** [comment]
@vshankar @chrisphoffman @batrick PTAL at priority as if confirmed this would have implications for fscrypt support in the kernel client [1].  The userspace fscrypt implementation seems to be somewhat lax around `NAME_MAX` so it may not be directly affected but would still need to be adjusted respectively:
https://github.com/ceph/ceph/blob/f3db497d115e2e99e015c18ae87b65cfdd166c17/src/client/FSCrypt.cc#L771-L792
(I didn't see where the long snap name suffix is appended though...)

[1] https://lore.kernel.org/ceph-devel/CAOi1vP96JuBi5BhTdiVeL491ziy_hwJUm79TqOb=iPaxck1eyg@mail.gmail.com/

**Comment by vshankar** [comment]
> @vshankar @chrisphoffman @batrick PTAL at priority

I will have to read back a bit on this. Will check.

File: src/mds/Server.h
Comment (vshankar) [reply src/mds/Server.h:463]: > This implies that `<INODE-NUMBER>` (in decimal, as it gets appended) can't be longer than 13 characters.

IIUC, this breaks when (say) using max_mds>=9, right?

Comment (idryomov) [reply src/mds/Server.h:463]: > IIUC, this breaks when (say) using max_mds>=9, right?

I hope it's not a question to me ;)  I'd appreciate if you elaborate on the relationship between `max_mds` and inode number space in general and how 9 was derived there in particular.

Comment (vshankar) [reply src/mds/Server.h:463]: Sorry for being vague. So, each (active) MDS would be assigned a partition of the inode space depending on the rank it holds. From src/mds/InoTable.cc:

```
void InoTable::reset_state()
{
  // use generic range. FIXME THIS IS CRAP
  free.clear();
  //#ifdef __LP64__
  uint64_t start = (uint64_t)(rank+1) << 40;
  uint64_t len = (uint64_t)1 << 40;
  //#else
  //# warning this looks like a 32-bit system, using small inode numbers.
  //  uint64_t start = (uint64_t)(mds->get_nodeid()+1) << 25;
  //  uint64_t end = ((uint64_t)(mds->get_nodeid()+2) << 25) - 1;
  //#endif
  free.insert(start, len);

  projected_free = free;
}
```

With 9 MDS ranks, the 9th rank is going to "fall off" the 13 digit limit (actually, even the 8th rank since the number of usable inodes is again 1<<40). So, this seems like a real problem/bug to me.

**Comment by chrisphoffman** [comment]
The libcephfs library does not have this problem as we do not encrypt snapshot names. This is because the manager creates subvolume snapshots and no key is to be present. I recommend we achieve feature parity by changing kernel functionality to match libcephfs. I've created a tracker for this proposed work here: https://tracker.ceph.com/issues/76257

**Comment by idryomov** [comment]
I'm not sure I'd refer to it as "achieving feature parity" as this would actually be removing a feature from the kernel client ;)

Is everyone convinced that encrypting snapshot names isn't useful in general (i.e. even outside of the subvolume snapshot framework)?  To me, a snapshot name isn't that much different from a file name -- both can contain potentially sensitive information.  Since file names are encrypted under fscrypt, it seems only logical to encrypt snapshot names as well.

When the manager creates a subvolume snapshot, does the snapshot name tend to be date+time/UUID/etc or is it more meaningful?

**Pull request merged**
